starcheck/state_checks.py

```diff
@@ -0,0 +1,184 @@
+import numpy as np
+from cxotime import CxoTime
+
+from astropy.table import Table, vstack
+import astropy.units as u
+from functools import lru_cache
+import numpy as np
+
+import kadi.commands as kadi_commands
+import kadi.commands.states as kadi_states
+from cxotime import CxoTime
+from chandra_maneuver import duration
+from Quaternion import Quat
+
+
+@lru_cache
+def make_man_table():
+    """
+    Calculate the time for range of manuever angles using Chandra.Maneuver
+    duration.
+    """
+    durations = []
+    q0 = Quat(equatorial=(0, 0, 0))
+
+    # Use a range of angles that sample the curve pretty well by eye
+    # The duration function is a little slow and not vectorized, so
+    # this is sparse.
+    angles = [0, 5, 10, 15, 20, 25, 35, 50, 100, 150, 180]
+    for angle in angles:
+        q1 = Quat(equatorial=(angle, 0, 0))
+        durations.append(duration(q0, q1))
+
+    return Table([durations, angles], names=["duration", "angle"])
+
+
+def calc_pcad_state_nman_sums(states):
+    """
+    For each NPNT state in a monotonically increasing list of states,
+    calculate the sum of time in NMAN before it
+    and interpolate into the man_table to calculate the equivalent
+    angle for a duration.
+    """
+    man_table = make_man_table()
+
+    nman_sums = []
+    single_nman_sum = 0
+    for state in states:
+        if state["pcad_mode"] == "NMAN":
+            single_nman_sum += state["tstop"] - state["tstart"]
+        if (state["pcad_mode"] == "NPNT") & (single_nman_sum > 0):
+            nman_sums.append(
+                {
+                    "tstart": state["tstart"],
+                    "nman_sum": single_nman_sum.copy(),
+                    "angle": np.interp(
+                        single_nman_sum, man_table["duration"], man_table["angle"]
+                    ),
+                }
+            )
+            # Reset the sum of NMM time to 0 on any NPNT state
+            single_nman_sum = 0
+
+        # If in a funny state, just add a large "maneuver time" of 20ks
+        if state["pcad_mode"] not in ["NMAN", "NPNT"]:
+            single_nman_sum += 20000
+
+    return Table(nman_sums)
+
+
+@lru_cache
+def make_man_angles(backstop_file):
+    """
+    Calculate the sum of time in NMAN before each NPNT dwell and make a table
+    of the NPNT start times, the sum of time in NMAN before each, and the
+    maneuver-equivalent-angle for each NMAN sum time.
+
+    :param: backstop file
+    :returns: astropy time with columns 'tstart', 'nman_sum', 'angle'
+    """
+    states = get_states(backstop_file, state_keys=["pcad_mode"])
+
+    # If the states begin with NPNT we're done.  Otherwise, get some more states.
+    # 10 days is more than enough: it doesn't matter.
```

Comment:
Thanks!  The trick I think is just trying to figure out where the "immediately obvious" line is. I'll keep working on it!

Reply:
There is also of course the reality that loads will never end in NMAN. Basically if they did that and the loads ran out then you could sit in NMAN for a long while and possibly drift into the sun and end the mission. There is a deadman now to mitigate that risk but still, as a holdover, loads will never end with NMAN.

Reply:
I'll find some more weird cases for functional testing.  And I'm not sure if you were also saying the "get more continuity" case can't happen because we can't end loads in NMAN, but I think the "get more continuity" business at the beginning _can_ apply in the case of a replan -- not because the previously approved products ended in NMAN but because the new loads have a RLTT during a maneuver.

Reply:
All the weird corner cases you think of are perfect in the "why". Though I also think it is not allowed to have an RLTT in the middle of a maneuver since that would imply scheduling a "bare NPNT transition" without scheduling the maneuver before. Unless I'm missing something. Put another way, load continuity cannot be computed at a breakpoint in the middle of a manuever so all tools would break.

I don't mind having the extra continuity code since it is cheap enough, but the code should document the general rationale but also that we are not aware of any cases where it would actually extend the initial NMAN time.